### PR TITLE
Improvements to SCIM group PATCH Operation performance

### PR DESCRIPTION
**Pull request by gayashanbc**
Resolves wso2/product-is#9120

### CODE REVIEW

**Comment by gayashanbc** [comment]
The build will pass once https://github.com/wso2/charon/pull/289 is merged.

**Review by emswbandara — APPROVED** [review]

**Comment by emswbandara** [comment]
@gayashanbc please resolve the conflicts

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 1]
```diff
@@ -2450,6 +2492,196 @@ public void updateGroup(Group oldGroup, Group newGroup) throws CharonException {
...
+                    String username = UserCoreUtil.addDomainToName(UserCoreUtil.removeDomainFromName(member),
+                            userStoreDomainForGroup);
+                    temporaryMembers.add(username);
+                }
+
+                deletedMembers.clear();
+                deletedMembers.addAll(temporaryMembers);
+            }
+
+            // Check for deleted members
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2573]: Shall we add a `.` at the end of the comment?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 2]
```diff
@@ -2450,6 +2492,196 @@ public void updateGroup(Group oldGroup, Group newGroup) throws CharonException {
...
+            Set<String> deletedMemberIdsFromUserstore =
+                    getMemberValuesFromUserstore(deletedMembers, userStoreDomainForGroup, newGroupName);
+
+            // Check for added members
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2577]: Shall we add a `.` at the end of the comment?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 3]
```diff
@@ -2450,6 +2492,196 @@ public void updateGroup(Group oldGroup, Group newGroup) throws CharonException {
...
+                validateUserIds(addedMemberIdsFromUserstore, newlyAddedMemberIds);
+            }
+
+            /*
+            Set thread local property to signal the downstream SCIMUserOperationListener
+            about the provisioning route.
+            */
+            SCIMCommonUtils.setThreadLocalIsManagedThroughSCIMEP(true);
+
+             /*
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2592]: Shall we fix alignment here?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 4]
```diff
@@ -2450,6 +2492,196 @@ public void updateGroup(Group oldGroup, Group newGroup) throws CharonException {
...
+             Update the group with added members and deleted members.
+              */
+            if (isNotEmpty(addedMembers) || isNotEmpty(deletedMembers)) {
+                carbonUM.updateUserListOfRoleWithID(newGroupName,
+                        deletedMemberIdsFromUserstore.toArray(new String[0]),
+                        addedMemberIdsFromUserstore.toArray(new String[0]));
+            }
+
+        } catch (UserStoreException e) {
+            log.error(e);
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2603]: Do we need to log and throw the error at the same time?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 5]
```diff
@@ -2450,6 +2492,196 @@ public void updateGroup(Group oldGroup, Group newGroup) throws CharonException {
...
+        if (userStoreDomainFromSP != null && !userStoreDomainFromSP.equalsIgnoreCase(oldGroupDomain)) {
+            throw new CharonException("Group :" + oldGroupName + "is not belong to user store " +
+                    userStoreDomainFromSP + "Hence group updating fail");
+        }
+
+        // If the updated group name does not contain a user store domain, it will be returned as PRIMARY.
+        String updatedGroupDomain = IdentityUtil.extractDomainFromName(newGroupName);
+
+        if (isPrimaryDomain(updatedGroupDomain) && !isPrimaryDomain(oldGroupDomain)) {
+             /*
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2664]: Shall we fix alignment issues in here?

Comment (gayashanbc) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/280/commits/36add81b74fe1bfc479adf2b8217e00af089b8f7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 4 above)
Comment (gayashanbc) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/280/commits/36add81b74fe1bfc479adf2b8217e00af089b8f7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 3 above)
Comment (gayashanbc) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/280/commits/36add81b74fe1bfc479adf2b8217e00af089b8f7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 2 above)
Comment (gayashanbc) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/280/commits/36add81b74fe1bfc479adf2b8217e00af089b8f7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 1 above)
Comment (gayashanbc) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/280/commits/36add81b74fe1bfc479adf2b8217e00af089b8f7

**Review by somindatommy — APPROVED** [review]

**Comment by gayashanbc** [comment]
Need to bump the kernel version before merging this PR once the new kernel is released

**Comment by IsuraD** [comment]
PR builder: https://github.com/wso2/product-is/actions/runs/358331772

**Comment by gayashanbc** [comment]
Merging as https://github.com/wso2/product-is/actions/runs/358331772 has passed

**Pull request merged**